litr/src/main/java/com/linkedin/android/litr/io/AudioRecordMediaSource.kt

```diff
@@ -0,0 +1,179 @@
+    }
+
+    // Compute the number of bytes per microsecond of audio.
+    private val bytesPerUs: Double by lazy {
+        (sampleRate * BYTES_PER_SAMPLE * channelCount) / 1000000.0
+    }
+
+    // The AudioRecord instance used to record the audio source.
+    private var audioRecord: AudioRecord? = null
+
+    // We track the amount of 16-bit PCM we have returned and calculate its duration, in order to
+    // know the presentation time of the next sample.
+    private var nextSampleTimeUs = 0L
+    private var sampleIncrementUs = 0L
+    private var totalDurationUs = 0L
+
+    private var isRecording = false
+
+    @Synchronized
+    fun start() {
+        createAudioRecord()
+
+        // Check to make sure the AudioRecord instance is initialized.
+        if (audioRecord?.state != AudioRecord.STATE_INITIALIZED) {
+            Log.e(TAG, "Error initializing AudioRecord")
+            throw MediaSourceException(
+                    MediaSourceException.Error.DATA_SOURCE,
+                    null,
+                    error("Error initializing AudioRecord: ${audioRecord?.state}"))
+        }
+
+        // Start recording.
+        audioRecord?.startRecording()
+        isRecording = true
+    }
+
+    private fun createAudioRecord() {
+        audioRecord = AudioRecord(
+                audioSource,
+                sampleRate,
+                channelConfig,
+                AUDIO_FORMAT,
+                bufferSize
+        )
+    }
+
+    @Synchronized
+    fun stop() {
+        if (isRecording) {
+            // Stop recording.
+            audioRecord?.stop()
+            isRecording = false
+        }
+    }
+
+    override fun getOrientationHint() = 0
+
+    override fun getTrackCount() = 1
+
+    override fun getTrackFormat(track: Int) = MediaFormat.createAudioFormat(
+            MimeType.AUDIO_RAW,
+            sampleRate,
+            channelCount
+    ).apply {
+        setInteger(MEDIA_FORMAT_PCM_KEY, AudioFormat.ENCODING_PCM_16BIT)
+    }
+
+    override fun selectTrack(track: Int) {
+        // Since we only support a single (audio) track, there is nothing to select.
+    }
+
+    override fun seekTo(position: Long, mode: Int) {
+        // We don't support seeking, since we're recording live.
+    }
+
+    override fun getSampleTrackIndex() = 0
+
+    override fun readSampleData(buffer: ByteBuffer, offset: Int): Int {
+        // If the recording has been stopped, then there are no more samples to be read.
+        if (!isRecording) {
+            return -1
+        }
+
+        // Read the next audio sample from the recorder. As per the AudioRecord documentation, we
+        // need to leave some buffer for the AudioRecord to continue to queue too. We also don't
+        // want to block waiting for more data, we'll take whatever is available.
+        val targetSize = (bufferSize / 2).coerceAtMost(buffer.capacity())
+        val readBytes = audioRecord?.read(buffer, targetSize, AudioRecord.READ_NON_BLOCKING) ?: -1
```

Comment:
To correctly satisfy `MediaSource` contract we should use `offset` value here and read into `buffer` at that value: https://developer.android.com/reference/android/media/MediaExtractor#readSampleData(java.nio.ByteBuffer,%20int) 
This is working because `offset` is 0, but it might not always be. This can be in a follow-up PR.

Reply:
Also, how is using `AudioRecord.READ_NON_BLOCKING` working? `readSampleData` is supposed to be a blocking call. 😅

Reply:
Thank you for explanation, now this makes sense. So basically we are telling `AudioRecord` not to wait until it fills its internal buffer and fill ours with whatever it has at the moment, synchronously. This makes a lot of sense and should work - as long as size is not -1 `AudioTrackTranscoder` should keep going.

Reply:
Yup, exactly.